packages/lib/gpc/src/gpcChecks.ts

```diff
@@ -538,24 +522,114 @@ export function checkVerifyClaimsForConfig(
     }
   }
+  // TODO(POD-P2): List membership claims
+}
+
+/**
+ * Picks the smallest available circuit in this family which can handle the
+ * size parameters of a desired configuration.
+ *
+ * @param circuitReq the circuit size requirements
+ * @returns the circuit description, or undefined if no circuit can handle
+ *   the required parameters.
+ * @throws Error if there are no circuits satisfying the given requirements.
+ */
+export function circuitDescForRequirements(
+  circuitReq: GPCRequirements
+): ProtoPODGPCCircuitDesc {
+  for (const circuitDesc of ProtoPODGPC.CIRCUIT_FAMILY) {
+    if (circuitDescMeetsRequirements(circuitDesc, circuitReq)) {
+      return circuitDesc;
+    }
+  }
+
+  throw new Error(
+    `There are no circuit parameters satisfying these requirements: ${circuitReq}`
+  );
+}
+
+/**
+ * Checks whether a described circuit can meet given GPC size requirements.
+ * This will be true if each of the circuit's parameters is greater than or
+ * equal to the required value.
+ *
+ * @param circuitDesc description of the circuit to check
+ * @param circuitReq the circuit size requirements
+ * @returns `true` if the circuit meets the requirements.
+ */
+export function circuitDescMeetsRequirements(
+  circuitDesc: ProtoPODGPCCircuitDesc,
+  circuitReq: GPCRequirements
+): boolean {
+  // Check tuple parameter compatibility.
+  // Indicate negative tuple requirement check if `requiredNumTuples` throws due
+  // to an invalid choice of parameters (i.e. `params.tupleArity < 2`).
+  const tupleCheck = ((): boolean => {
+    try {
+      // The circuit description should have enough tuples of arity `tupleArity` to
+      // cover all input tuples.
+      return (
+        circuitDesc.maxTuples >=
+        circuitReq.tupleArities
+          .map((arity) => requiredNumTuples(circuitDesc.tupleArity, arity))
+          .reduce((sum, requiredNum) => sum + requiredNum, 0)
+      );
+    } catch (_) {
+      return false;
+    }
+  })();
+  return (
+    tupleCheck &&
+    circuitDesc.maxObjects >= circuitReq.nObjects &&
+    circuitDesc.maxEntries >= circuitReq.nEntries &&
+    circuitDesc.merkleMaxDepth >= circuitReq.merkleMaxDepth &&
+    circuitDesc.maxLists >= circuitReq.nListElements.length &&
+    // The circuit description should be able to contain the largest of the lists.
+    circuitDesc.maxListElements >= Math.max(...circuitReq.nListElements, 0)
+  );
+}
+
+/**
+ * Calculates the merged set of GPC size requirements meeting the unified
+ * (maximum) requirements of both inputs.
+ *
+ * @param rs1 first set of required sizes
+ * @param rs2 second set of required sizes
+ * @returns unified (maximum) sizes
+ */
+
+export function mergeRequirements(
+  rs1: GPCRequirements,
+  rs2: GPCRequirements
+): GPCRequirements {
+  return GPCRequirements(
+    Math.max(rs1.nObjects, rs2.nObjects),
+    Math.max(rs1.nEntries, rs2.nEntries),
+    Math.max(rs1.merkleMaxDepth, rs2.merkleMaxDepth),
+    rs1.nListElements.map((nElements, i) =>
```

Comment:
I'm not sure if this will be the right approach when the config handling is implemented.
In particular, merging is usually between config-based requirements and input-based requirements.
The config and input might not both know about tuples at all.  If they do, it might be complicated to get them to agree about the order in which they appear in these lists.

The latter might be clearer if they were in an object by name, though it's not clear to me whether tuples and lists will always need to be assigned a name.  The former you can figure out when the config format gets settled, but it might suggest these lines should be appending rather than doing an elementwise max.

Reply:
Pondering over this one, I'm wondering whether it makes sense to put any of the list or tuple stuff in `GPCProofConfig` vs just having it all in `GPCProofInputs`. At the moment, the only use for tuples is for list membership checks, and from what I gather list membership checks are more of a proof-specific feature, thus they ought to go in the proof input rather than the proof config. Unlike the Merkle tree depth, there isn't really any reasonable way to estimate the sizes of lists, since they are independent of the other inputs, and we agreed that putting the list size in the config was a no-go, so including list memberships in the proof config wouldn't serve any purpose. This would imply the tuple and list requirements should be trivial for proof configurations. Does this make sense, or am I missing something?

Reply:
I think ProofConfig definitely needs to specify all of the constraints of the proof, while inputs contains only data.  So if anything lists & tuples would go entirely in config, rather than entirely inputs.  Actually I think tuples probably do go entirely in config, but the contents of the list is something I'd argue belongs in inputs, similar to the contents of PODs.  (I think that's a bit questionable, because there will likely be some cases where the lists will always be the same, and other cases where they will vary.)  List constraint in config with list contents in input implies that lists definitely will need names to map between config and inputs, same as PODs.  Tuples don't seem to need names yet, since they might be entirely implied by the list config, as you point out.

Regardless, without restructuring the code, there has to be some value in the Requirements object generated from config and inputs in order to merge.  I'd expect the config requirements to be "1 list of size >-1" while the inputs specify a larger list size.  The use of gpcBindConfig() requires generating some minimum set of requirements without input.  I'd expect it to be able to pick a circuit with 1 list of the smallest available size.  That's similar to what it does with PODs and Merkle depth.  If the caller needs to estimate a larger size, they can do it by picking a circuit explicitly.  We could allow them to pass their own requirements object too, which may make sense, but was something I didn't implement yet to keep the exposed complexity relatively low.  It's just as easy for someone to run a single test case of gpcProve() and save the circuit identifier which comes out after it's calculated.

Regardless, I think this might be overthinking for this PR.  I'd expect you to have to iterate on these details as you implement the real config + input handling, and we'll be able to see how it all works together better when that's done.  I think it's okay if you land this PR with this code as-is, or with some guess at how it'll work, and update it later.  I think splitting your efforts into smaller steps is great, but you don't need to perfectly predict what's coming in the next PR.